_posts/2023-11-17-transactional-outbox.md

```diff
@@ -0,0 +1,295 @@
+The system thus ends up in an inconsistent state, with the event bus reflecting a change that the database does not.
+
+When integrating an event bus or any other external service within an application in a distributed system, there are two naive approaches one might consider. 
+Each approach has its own set of challenges and potential pitfalls:
+ 
+## 1. Execute Business Logic Within the Transaction and Publish Events After Committing the Transaction
+In this approach, the application first executes all the required business logic within a database transaction. 
+After this transaction is successfully committed, it then publishes the relevant events to the event bus.
+
+![commit-publish](/assets/article_images/commit-publish.png)
+
+**Challenges:**
+
+* **Risk of Event Loss:** If the application crashes or encounters an issue between the transaction commit and the event publication, the event may never get published. 
+This leads to a situation where the database state has changed, but the corresponding event informing other services of this change is lost.
+* **Inconsistency in Distributed Systems**: Other services relying on these events for triggering their own processes will not be aware of the changes, leading to data inconsistency across the system.
+* **Complex Recovery Mechanisms**: Implementing a mechanism to check for such failures and recover from them can be complex and error-prone.
+
+This approach can be suitable for situations where the events being published are not critical to the core business logic or data integrity of the application. 
+An example of this might be events used for monitoring or logging purposes. 
+In these cases, the loss of some events due to system failures or crashes might be considered an acceptable risk.
+
+## 2. Execute Business Logic and Publish Events Within the Transaction
+In this approach, the application attempts to execute business logic and publish events to the event bus, all within the same database transaction. 
+
+
+![publish-commit](/assets/article_images/publish-commit.png)
+
+**Challenges:**
+* **Event Publication Preceding Transaction Commit:** When events are published to the event bus as part of the database transaction, there's a risk that the transaction might not successfully commit even after the event has been sent out. 
+This situation can occur due to various reasons, such as issues in completing subsequent operations within the same transaction or a late-occurring failure in the database.
+* **Inconsistent State Risks:** If the transaction fails to commit after the event is published, it leads to a critical inconsistency. 
+The event bus carries an event that suggests a change or action that, in reality, the database did not finalize. This discrepancy can lead to serious data integrity issues, as other services responding to the event may act on data that doesn't accurately reflect the current state of the system.
```

Comment:
an event rollback on commit failure would also not make sense, because the event may already have triggered some other process?

Reply:
That's correct. It will the complex error handling, which is not feasible.